extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
-                    verifyCodeFlowAccessToken(vertxContext, request, resolvedContext));
+        Uni<AccessToken> codeAccessTokenUni = verifyCodeFlowAccessTokenUni(vertxContext, request, resolvedContext);
+
+        return codeAccessTokenUni.onItem().transformToUni(
+                new Function<AccessToken, Uni<? extends SecurityIdentity>>() {
+                    @Override
+                    public Uni<SecurityIdentity> apply(AccessToken codeAccessToken) {
+                        return validateTokenWithOidcServer(vertxContext, request, resolvedContext, codeAccessToken);
+                    }
+                });
+    }
+
+    @SuppressWarnings("deprecation")
+    private Uni<SecurityIdentity> validateTokenWithOidcServer(RoutingContext vertxContext, TokenAuthenticationRequest request,
+            TenantConfigContext resolvedContext, AccessToken codeAccessToken) {
+
+        if (codeAccessToken != null) {
+            vertxContext.put("code_flow_access_token_result", codeAccessToken);
         }
 
-        final JsonObject userInfo = resolvedContext.oidcConfig.authentication.isUserInfoRequired()
-                ? getUserInfo(vertxContext, request, resolvedContext)
-                : null;
+        Uni<JsonObject> userInfo = getUserInfoUni(vertxContext, request, resolvedContext);
 
-        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
-            @Override
-            public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+        return userInfo.onItem().transformToUni(
+                new Function<JsonObject, Uni<? extends SecurityIdentity>>() {
+                    @Override
+                    public Uni<SecurityIdentity> apply(JsonObject userInfo) {
+                        return createSecurityIdentityWithOidcServerUni(vertxContext, request, resolvedContext, userInfo);
+                    }
+                });
+    }
+
+    private Uni<SecurityIdentity> createSecurityIdentityWithOidcServerUni(RoutingContext vertxContext,
+            TokenAuthenticationRequest request, TenantConfigContext resolvedContext, final JsonObject userInfo) {
 
-                resolvedContext.auth.decodeToken(request.getToken().getToken(),
-                        new Handler<AsyncResult<AccessToken>>() {
+        final TokenCredential tokenCred = request.getToken();
+        if (tokenCred instanceof AccessTokenCredential && ((AccessTokenCredential) tokenCred).isOpaque()) {
+            // remote introspection is required, a blocking call
+            if (BlockingOperationControl.isBlockingAllowed()) {
+                return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                    @Override
+                    public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                        createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext, userInfo);
```

Comment:
even if blocking is allowed, it may be not called on the same thread. This method is called on the subscription thread, not the assembly thread.

Reply:
@cescoffier I'm just trying to follow the same pattern as Stuart does [here](https://github.com/quarkusio/quarkus/blob/master/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusIdentityProviderManagerImpl.java#L54). I can remove these `BlockingOperationContext` checks ?

Reply:
And yes, it will be gone once that refresh token check becomes Uni async friendy, I'm actually nearly finished in another branch

Reply:
Except that in Stuart's code, the check is done at subscription time. You are doing it at assembly time and hope for the best are subscription:

```java
// This is called at assembly time
Uni.createFrom().emitter(e -> {
   // This is called at subscription time
});
```

The thread used at assembly and subscription time may be different. So if assembly time can block, it does not means you can block at subscription time (or the opposite).

Reply:
Hey @cescoffier I'm on it, I'm learning something new here and at this rate I may qualify as a junior Mutiny bug fixer and start helping @jponge :-)